### PR TITLE
CalendarWrapper: setdaystate, sizerecttomin, hittest

### PR DESCRIPTION
**Pull request by VVillwin**

### CODE REVIEW

File: pywinauto/controls/common_controls.py [hunk 1]
```diff
@@ -3388,6 +3388,58 @@ def set_view(self, viewType):
...
+        res = self.send_message(win32defines.MCM_SETDAYSTATE, len(day_states), remote_mem)
+        del remote_mem
+
+        if res == 0:
+            raise RuntimeError('Failed to set the day states in Calendar')
+
+        return res
+
+    #----------------------------------------------------------------
+    def minimize_calendar_rectangle(self, left, top, right, bottom):
```
Comment (airelil) [inline pywinauto/controls/common_controls.py]: I would call it ```calc_min_rectangle``` (as I understand there is no 'calendar' in other methods names)

File: pywinauto/unittests/test_Calendar.py [hunk 2]
```diff
@@ -86,11 +88,102 @@ def test_can_set_calendars_view_into_century(self):
...
 
+    def get_expected_minimized_rectangle(self):
+        expected_rect = win32structures.RECT()
+        expected_rect.left = 0
+        expected_rect.top = 0
+        expected_rect.right = 162
+        expected_rect.bottom = 160
+        return expected_rect
+
+    def set_calendar_state_to_display_day_states(self):
```
Comment (airelil) [inline pywinauto/unittests/test_Calendar.py]: Internal helper functions should be started with an underscore ```_set_calendar_stUsinate_to_display_day_states```

File: pywinauto/unittests/test_Calendar.py [hunk 3]
```diff
@@ -86,11 +88,102 @@ def test_can_set_calendars_view_into_century(self):
...
         self.assertEqual(systemTime.wYear, now.year)
         self.assertEqual(systemTime.wMonth, now.month)
         self.assertEqual(systemTime.wDay, now.day)
 
+    def get_expected_minimized_rectangle(self):
+        expected_rect = win32structures.RECT()
+        expected_rect.left = 0
+        expected_rect.top = 0
+        expected_rect.right = 162
+        expected_rect.bottom = 160
```
Comment (airelil) [inline pywinauto/unittests/test_Calendar.py]: Using hard coded coordinates here and in previous tests with ```do_hit_test``` isn't reliable. The control size and coordinates change under different windows themes. We already have this problem in other tests. Maybe you should try to get the main control dimensions before testing and use them as the basis for the rest?

Comment (VVillwin) [inline pywinauto/unittests/test_Calendar.py]: Would it be suitable to call Rectangle() for Calendar's wrapper and calculate sizes?
Smth like that:
rect = self.app['Common Simple Controls']['Calendar'].WrapperObject().Rectangle()
#rect.width(), rect.height()

File: pywinauto/controls/common_controls.py [hunk 4]
```diff
@@ -3388,6 +3388,58 @@ def set_view(self, viewType):
         if res == 0:
             raise RuntimeError('Failed to set view in Calendar')
 
+    #----------------------------------------------------------------
+    def set_day_states(self, month_states):
+        """Sets the day states for all months that are currently visible within a month calendar control"""
```
Comment (vasily-v-ryabov) [inline pywinauto/controls/common_controls.py]: Phrase `within a month calendar control` is obsolete and it makes the doc string too long.

File: pywinauto/controls/common_controls.py [hunk 5]
```diff
@@ -3388,6 +3388,58 @@ def set_view(self, viewType):
         if res == 0:
             raise RuntimeError('Failed to set view in Calendar')
 
+    #----------------------------------------------------------------
+    def set_day_states(self, month_states):
+        """Sets the day states for all months that are currently visible within a month calendar control"""
+        remote_mem = RemoteMemoryBlock(self)
+        day_states = (ctypes.c_uint32 * len(month_states))(*month_states)
```
Comment (vasily-v-ryabov) [inline pywinauto/controls/common_controls.py]: I see `typedef DWORD MONTHDAYSTATE` in [the MSDN docs](https://msdn.microsoft.com/en-us/library/windows/desktop/bb760915(v=vs.85).aspx). `DWORD` is defintely smaller than `ctypes.c_uint32`. You may check it by `ctypes.sizeof()` function. How about using `from ctypes import wintypes; wintypes.DWORD` instead?

File: pywinauto/unittests/test_Calendar.py [hunk 6]
```diff
@@ -86,11 +88,102 @@ def test_can_set_calendars_view_into_century(self):
         self.calendar.set_view(win32defines.MCMV_CENTURY)
         self.assertEqual(self.calendar.get_view(), win32defines.MCMV_CENTURY)
 
+    def test_can_set_day_state(self):
+        month_states = [self.NO_HOLIDAYS_IN_MONTH, self.NO_HOLIDAYS_IN_MONTH, self.NO_HOLIDAYS_IN_MONTH]
```
Comment (vasily-v-ryabov) [inline pywinauto/unittests/test_Calendar.py]: Did you try another values?

File: pywinauto/unittests/test_Calendar.py (hunk 3 above)
Comment (vasily-v-ryabov) [inline pywinauto/unittests/test_Calendar.py]: Yes, something like that. By the way, `WrapperObject()` call is not necessary in a production code. The control will be found automatically. You need it just for debugging.

File: pywinauto/unittests/test_Calendar.py (hunk 6 above)
Comment (VVillwin) [inline pywinauto/unittests/test_Calendar.py]: Yes. I've not found any other cases which may lead to res equals to 0

**Comment by vasily-v-ryabov** [comment]
The tests don't work on my Win7 laptop (En-US locale):
```
======================================================================
FAIL: test_can_determine_calendar_background_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 129, in test_can_determine_calendar_background_is_hit
    self.assertEquals(win32defines.MCHT_CALENDARBK, res)
AssertionError: 131072 != 0L

======================================================================
FAIL: test_can_determine_calendar_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 121, in test_can_determine_calendar_is_hit
    self.assertEquals(win32defines.MCHT_CALENDAR, res)
AssertionError: 131072 != 0L

======================================================================
FAIL: test_can_determine_title_background_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 181, in test_can_determine_title_background_is_hit
    self.assertEquals(win32defines.MCHT_TITLEBK, res)
AssertionError: 65536 != 33619971L

======================================================================
FAIL: test_can_determine_title_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 173, in test_can_determine_title_is_hit
    self.assertEquals(win32defines.MCHT_TITLE, res)
AssertionError: 65536 != 33619971L

======================================================================
FAIL: test_can_determine_week_number_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 214, in test_can_determine_week_number_is_hit
    self.assertEquals(win32defines.MCHT_CALENDARWEEKNUM, res)
AssertionError: 131075 != 0L

----------------------------------------------------------------------
Ran 30 tests in 4.863s

FAILED (failures=5)
```

**Comment by VVillwin** [comment]
Please try now.
Had the tests failed before? Or just this time?

**Comment by vasily-v-ryabov** [comment]
The same failures. Never checked on the laptop before. But need to fix that.
OS language is English, formats are Russian (first tab of settings), location is US (second tab of settings in Control panel).
os_locale

**Comment by VVillwin** [comment]
The provided settings are the same for me. But I have Windows 8.1 and probably different screen.
Actually I have no idea how to fix it.

I've changed some coordinates to be more in the center of the regions. But I am not sure if it helps.

**Comment by vasily-v-ryabov** [comment]
The rest of failures for now (it might be 64-bit specific errors on py2.7 x64):
```
FAIL: test_can_determine_calendar_background_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 129, in test_can_determine_calendar_background_is_hit
    self.assertEquals(win32defines.MCHT_CALENDARBK, res)
AssertionError: 131072 != 33685505L

======================================================================
FAIL: test_can_determine_calendar_is_hit (__main__.CalendarWrapperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\pywinauto\unittests\test_Calendar.py", line 121, in test_can_determine_calendar_is_hit
    self.assertEquals(win32defines.MCHT_CALENDAR, res)
AssertionError: 131072 != 33685505L

----------------------------------------------------------------------
Ran 30 tests in 5.397s

FAILED (failures=2)
```

**Comment by vasily-v-ryabov** [comment]
By the way, you need to merge recent changes from the upstream repo to your fork. It will make the PR auto-mergable.

File: pywinauto/controls/common_controls.py [hunk 7]
```diff
@@ -3496,7 +3496,7 @@ def calc_min_rectangle(self, left, top, right, bottom):
         return minimized_rect
 
     #----------------------------------------------------------------
-    def do_hit_test(self, x, y):
+    def test_hit(self, x, y):
```
Comment (vasily-v-ryabov) [inline pywinauto/controls/common_controls.py]: -> `hit_test`

**Comment by vasily-v-ryabov** [comment]
Please merge recent commits from the upstream/UIA to make the PR auto mergeable. Local tests run seems OK.

**Pull request merged**